[Dynamo] Backend registration with ``entry_points``

yaoyaoding opened the pull request:
Fixes #91824

This PR add a new dynamo backend registration mechanism through ``entry_points``. The ``entry_points`` of a package is provides a way for the package to reigster a plugin for another one. 

The docs of the new mechanism:
![image](https://user-images.githubusercontent.com/23381083/216133221-18cf18e2-6ad6-4cf7-8da2-9b9b883389c8.png)
(the typo '...named "my_backend" that has been..." has been fixed to '...named "my_compiler" that has been...')

# Discussion

## About the test
I did not add a test for this PR as it is hard either to install a fack package during a test or manually hack the entry points function by replacing it with a fake one. I have tested this PR offline with the hidet compiler and it works fine. Please let me know if you have any good idea to test this PR.

## About the dependency of ``importlib_metadata``
This PR will add a dependency ``importlib_metadata`` for the python < 3.10 because the modern usage of ``importlib`` gets stable at this python version (see the documentation of the importlib package [here](https://docs.python.org/3/library/importlib.html)).  For python < 3.10, the package ``importlib_metadata`` implements the feature of ``importlib``. The current PR will hint the user to install this ``importlib_metata`` if their python version < 3.10. 

## About the name and docs
Please let me know how do you think the name ``torch_dynamo_backend`` as the entry point group name and the documentation of this registration mechanism.


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

yaoyaoding commented:
@ezyang @voznesenskym @dzhulgakov Could you have a look at this PR? Thanks.

malfet reviewed (approved):
Looks fine, but please turn else into it's own routine and make it memoizable

jansel commented:
This PR will conflict with https://github.com/pytorch/pytorch/pull/93389 which moves/refactors this code.

jansel commented on torch/_dynamo/eval_frame.py:
```diff
@@ -364,7 +364,28 @@ def lookup_backend(compiler_fn):
+                try:
+                    from importlib_metadata import entry_points
+                except ImportError:
+                    raise RuntimeError(
```
I'd suggest just silently ignoring this error rather than raising.  My thinking is: if the user doesn't have `entry_points` installed, then clearly there aren't any backends installed in it and "unknown backend" is a better error message.

IMO the "please install `entry_points`" error should come from when the backend tries to register itself (which happens first), not when we try to look it up.

dzhulgakov replied:
How about you just customize the error below, i.e. have something like "Cannot recognize backend: {}. Plugin mechanism for backend is disabled because importlib-metadata is not installed. Please run `pip3 install importlib-metadata` if you believe you installed external Dynamo plugin" (but only if the backend wasn't found)

jansel replied:
Shouldn't `pip install my-awesome-backend` depend on `entry_points` and trigger it to get installed automatically?

yaoyaoding replied:
@dzhulgakov Thanks for the suggestion! I agree that it is better to mention the backend name when raise the error as you said. Will update the error message.

@jansel Sorry for the confusion. The ``entry_points`` is not a package but a functionality in distributing a package. Thus, ``pip install dynamo-backend`` does not depends on ``entry_points``. Its interface gets stable in 3.10 and the querying function is implemented in a standard library ``importlib`` (see its [docs](https://docs.python.org/3/library/importlib.html)). For python < 3.10, we need package ``importlib_metadata`` to access this "future" standard library.

jansel replied:
I am suggesting that backend packages should contain something like this in their `setup.py`:
```py
install_requires = [...]
if sys.version_info < (3, 10):
    install_requires.append("importlib_metadata")

setup(
    ...
    install_requires=install_requires,
    ...
)
```

yaoyaoding replied:
> This PR will conflict with #93389 which moves/refactors this code.

Should I wait for that PR to be merged or rebase to that PR to continue?

yaoyaoding replied:
> I am suggesting that backend packages should contain something like this in their `setup.py`:
> 
> ```python
> install_requires = [...]
> if sys.version_info < (3, 10):
>     install_requires.append("importlib_metadata")
> 
> setup(
>     ...
>     install_requires=install_requires,
>     ...
> )
> ```
> 
> And shouldn't need to ask the user to manually install it. That way if `importlib_metadata` is missing, we know there are no backends using it installed.

I see. My concern is that the backend has no actual dependency on the ``importlib_metadata`` package as they might never use it in their codebase. 

I find that PyTorch 2.0 is going to drop the support for python 3.7. That's a good news because python introduced the ``importlib.metadata`` since python 3.8, though it is in provisional status (see the [docs](https://docs.python.org/3.8/library/importlib.metadata.html) here). Comparing with requiring the backend packages to add a new dependency, I prefer to use ``importlib.metadata`` (the usage will be slightly different for python 3.8, 3.9 vs 3.10 and onwards).

jansel replied:
> I see. My concern is that the backend has no actual dependency on the `importlib_metadata` package as they might never use it in their codebase.

To be useable by PyTorch+importlib_metadata, the backend packages will need to populate the necessary metadata in their setup.py.  IMO part of that metadata population should involve adding the needed dependency.  If the backends don't do specific work to define `entry_points()` it will just be empty.

yaoyaoding replied:
Adding the entry_points information does not require the existence of package ``importlib_metadata``. There are different ways to access the ``entry_points`` information in the current python enviroment. In very old time, people use package ``pkg_resources``. Now people use ``importlib_metadata`` (python < 3.8) and ``importlib.metadata`` (python >= 3.8). 

Since PyTorch 2.0 is going to only support python >= 3.8, I believe now there is no problem on the dependency as we can only use the standard python package ``importlib``. I have updated the PR to only use ``importlib`` and now this PR does not introduce any new dependency.

yaoyaoding replied:
Hi @jansel, could you have a look at this PR again? Please let me know if you have any suggestion, thank you!

dzhulgakov commented on torch/_dynamo/eval_frame.py:
```diff
@@ -364,7 +364,28 @@ def lookup_backend(compiler_fn):
+        if compiler_fn in BACKENDS:
+            compiler_fn = BACKENDS[compiler_fn]
+        else:
+            # See https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/#using-package-metadata
+            # for more information on using entry_points to provide plugins for python packages.
+            if sys.version_info < (3, 10):
+                try:
+                    from importlib_metadata import entry_points
+                except ImportError:
+                    raise RuntimeError(
+                        "For Python < 3.10, please install importlib-metadata via\n"
+                        "  pip3 install importlib-metadata"
+                    )
+            else:
+                from importlib.metadata import entry_points
+            backend_name = compiler_fn
+            backend_eps = entry_points(group="torch_dynamo_backends")
+            if backend_name in backend_eps.names:
+                compiler_fn = backend_eps[backend_name].load()
+            else:
+                raise RuntimeError("Cannot recognize backend: {}".format(compiler_fn))
```
on that note: on success, you can put the looked-up backend into `BACKENDS` so the next execution is way faster

yaoyaoding replied:
Added the memoization in the same function.

malfet replied:
Sorry, I don't see where you did it?

malfet commented on torch/_dynamo/eval_frame.py:
```diff
@@ -364,7 +365,25 @@ def lookup_backend(compiler_fn):
+                if backend_name in backend_eps.names:
+                    compiler_fn = backend_eps[backend_name].load()
+            if compiler_fn is None:
+                raise RuntimeError("Cannot recognize backend: {}".format(backend_name))
```
```suggestion
                raise RuntimeError(f"Cannot recognize backend: {backend_name}")
```

malfet commented on torch/_dynamo/eval_frame.py:
```diff
@@ -364,7 +365,25 @@ def lookup_backend(compiler_fn):
         from .optimizations import BACKENDS
 
-        compiler_fn = BACKENDS[compiler_fn]
+        backend_name = compiler_fn
+        if backend_name not in BACKENDS:
+            compiler_fn = None
+            group_name = "torch_dynamo_backends"
+            if sys.version_info < (3, 10):
+                backend_eps = entry_points()
+                eps = [ep for ep in backend_eps[group_name] if ep.name == backend_name]
+                if len(eps) > 0:
+                    compiler_fn = eps[0].load()
+            else:
+                backend_eps = entry_points(group=group_name)
+                if backend_name in backend_eps.names:
+                    compiler_fn = backend_eps[backend_name].load()
+            if compiler_fn is None:
+                raise RuntimeError("Cannot recognize backend: {}".format(backend_name))
+            BACKENDS[backend_name] = compiler_fn
+
+        compiler_fn = BACKENDS[backend_name]
```
Please move all the logic to memoizable `find_backend_` function
```suggestion
        backend_name = compiler_fn
        compiler_fn = find_backend_(backend_name)
        if compiler_fn is None:
            raise RuntimeError("Cannot recognize backend: {}".format(backend_name))
```

yaoyaoding replied:
Hi @malfet, I memorize the found backend to `BACKENDS` at Line 383 (suggested by @dzhulgakov), thus we will not search it in ``entry_points`` in subsequent usage. I feel it slightly better to store both the predefined and external (i.e., through ``entry_points``) backends in the same place. 

Of course, we can also put all the logic (lookup `BACKENDS` and `entry_points`) in another function and cache it through `functools.lru_cache`. If you prefer this way, let me know and I will update.

yaoyaoding replied:
I have seen the refactored registration in the main branch, and seems it is better to use a new function to lookup the entry points backend. I will let you know when I finish rebase.

malfet reviewed (changes requested):
This looks like a much cleaner change now, thank you for working on it.

But please rebase your PR (as it conflicts with trunk right now) and move the logic of searching for a backend via extensions to its own memoizable function (and I see the benefit of not preloading all the extensions at the beginning)

yaoyaoding commented on torch/_dynamo/eval_frame.py:
```diff
@@ -364,7 +365,25 @@ def lookup_backend(compiler_fn):
+                    compiler_fn = backend_eps[backend_name].load()
+            if compiler_fn is None:
+                raise RuntimeError("Cannot recognize backend: {}".format(backend_name))
+            BACKENDS[backend_name] = compiler_fn
```
This Line is used to memorize the found backend through `entry_points`.

yaoyaoding committed:
rebase

yaoyaoding committed:
fix

yaoyaoding commented:
Hi @malfet and @jansel, I have rebased this PR to the current main branch of pytorch repo, and updated the PR according to your suggestions. Could you have a look again? Thanks!

yaoyaoding committed:
lint

yaoyaoding commented:
Kindly ping @malfet as today is the release branch cut.

dzhulgakov commented:
@pytorchbot merge

yaoyaoding commented:
Hi @malfet and @jansel, I checked the log of failed ci job (macos-12-py3 / test) and found the the following error
```
Traceback (most recent call last):
  File "inductor/test_perf.py", line 16, in <module>
    from torch.testing._internal.inductor_utils import HAS_CUDA
  File "/Users/runner/work/_temp/conda_environment_4171205152/lib/python3.8/site-packages/torch/testing/_internal/inductor_utils.py", line 13, in <module>
    CppCodeCache.load("")
  File "/Users/runner/work/_temp/conda_environment_4171205152/lib/python3.8/site-packages/torch/_inductor/codecache.py", line 494, in load
    from filelock import FileLock
ModuleNotFoundError: No module named 'filelock'
```
Do you have any idea to fix it? Seems like this error is not caused by this PR.

malfet commented:
@yaoyaoding looks like a CI flake, feel free to ignore
